src/sentry/workflow_engine/processors/action.py

```diff
@@ -121,6 +121,36 @@ def update_workflow_action_group_statuses(
     )
 
 
+def deduplicate_actions(
+    actions_queryset: BaseQuerySet[Action], action_to_workflow_ids: dict[int, int]
+) -> BaseQuerySet[Action]:
+    """
+    Deduplicates actions based on their handler's dedup_key method.
+    Returns a filtered and annotated queryset with duplicates removed.
+    When duplicates are found, keeps the action with the lowest ID.
+    """
+    # Sort actions by ID to ensure consistent ordering for deduplication
+    # We sort reverse since we rely on overwriting the dedup_key_to_action_id dict
+    sorted_actions = sorted(actions_queryset, key=lambda a: a.id, reverse=True)
+
+    dedup_key_to_action_id: dict[str, int] = defaultdict(int)
+
+    for action in sorted_actions:
+        dedup_key = action.get_dedup_key()
+
+        dedup_key_to_action_id[dedup_key] = action.id
+
+    # Create workflow_id annotation cases for final actions
+    workflow_id_cases = [
+        When(id=action_id, then=Value(action_to_workflow_ids[action_id]))
+        for action_id in dedup_key_to_action_id.values()
+    ]
+
+    return Action.objects.filter(id__in=dedup_key_to_action_id.values()).annotate(
+        workflow_id=Case(*workflow_id_cases, output_field=models.IntegerField()),
+    )
```

Comment:
this will require us to do 2 queries for actions, the first to get the base query of actions to build the dictionary, then the other here to get the actions + annotations.

instead, we could pass in the decorated actions and use the dictionary to get back down to a single query:
```suggestion
def deduplicate_actions(
    actions_queryset: BaseQuerySet[Action],  # decorated with the workflow_ids
    action_to_workflow_ids: dict[int, int],
) -> BaseQuerySet[Action]:
    """
    Deduplicates actions based on their handler's dedup_key method.
    Returns a de-duplicated list of actions.
    """    
    dedup_key_to_action_id: dict[str, Action] = defaultdict()
    
    for action in sorted_actions:
        dedup_key = action.get_dedup_key()
        dedup_key_to_action_id[dedup_key] = action

    return set(dedupe_key_to_action_id.values())
```

What's nice about this is, it's only doing what the method name says -- it's not adding additional annotations or anything, it's just using the queryset of actions to determine a de-duplicated set of actions.